### ISSUE TITLE
Inconsistent tsc behavior on TS 5.4

### ISSUE THREAD

**Issue by olegafx** [microsoft/TypeScript#58175]
### 🔎 Search Terms

"tsc inconsistent behavior", "tsc 0 errors", "ts 5.4 bug", "5.4.0-dev.20240219"

### 🕗 Version & Regression Information

I wanted to update TS in our project from `v5.3.3` to `v5.4.3`. And I noticed several weird things:
- When I was switching TS version in VS Code, it wasn't showing any errors. But `tsc --noEmit` was showing 2 errors. I thought that VS Code was just not ready to support `v5.4.3` yet. But the new version comes with `v5.4.3` and it still doesn't work correctly.
- I run `tsc --noEmit`. It shows `2 errors`.  If I open a problematic file and put `// @ts-expect-error`, it shows only `1 error`. But if I remove that line, it instantly shows `0 errors`.

This changed between versions `5.4.0-dev.20240218` and `5.4.0-dev.20240219`.

`5.4.0-dev.20240219` is the first version where I see this problem. I checked `@next` version which is `5.5.0-dev.20240412` and the error is still there.

Unfortunately, I wasn't able to make a reproducible example due complexity of the code. But both errors are about `unknown` type where it should be detected just fine.

### 🙁 Actual behavior

- It doesn't show errors in VS Code
- It shows errors in a console. But after applying `// @ts-expect-error` and removing that line, it shows `0 errors`.

### 🙂 Expected behavior

- Errors in VS Code are detected
- It shows errors in a console. After applying `// @ts-expect-error` and removing that line, it shows all previous errors.

**Comment by RyanCavanaugh** [comment]
There's really not much we can do without a way to reproduce the problem, unfortunately

**Comment by olegafx** [comment]
Ok, I was able to reproduce it. Here is a [link](https://stackblitz.com/edit/vitejs-vite-dy48pa?file=src%2Ffail%2Ffail.tsx) to StackBlitz. There are 2 main cases:

1. If you'll open `fail.tsx`, you won't see any errors in the editor. But if you will run `yarn tsc --noEmit` in the terminal, you'll see `src/fail/fail.tsx:18:22 - error TS2322: Type 'unknown[]' is not assignable to type 'FailedItem[]'`.
2. Start `tsc` in watch mode: `yarn tsc --noEmit --watch`. Open `fail.tsx` and add `// @ts-expect-error` in any place of the code. Remove this comment. It will show `0 errors` in the console.

Also you could notice that there is `import '@total-typescript/ts-reset/is-array';` in `reset.d.ts`. The only thing it's doing is changing `any` to `unknown` for `Array.isArray`. [Source code](https://github.com/total-typescript/ts-reset/blob/main/src/entrypoints/is-array.d.ts).

React version doesn't really matter. It works the same on `v16` and `v18`.

**Comment by DanielRosenwasser** [comment]
For what it's worth, the only thing that was merged in [between those dates](https://github.com/microsoft/TypeScript/pulls?q=is%3Apr+merged%3A2024-02-17..2024-02-19+is%3Aclosed) was https://github.com/microsoft/TypeScript/pull/57122

**Comment by ahejlsberg** [comment]
Even shorter repro:

```ts
type TX<T extends any[] & object> = T["length"];  // Comment this out and error goes away
type T0<T extends any[] & object> = T;
type T1 = T0<string[]>;  // Error!
```

**Comment by olegafx** [comment]
Just update our project to TS 5.5. And everything works great! Thanks a lot, guys! 💪🏻
